From 62821d2db6a1d60bc00c22d4da021999772bd2e4 Mon Sep 17 00:00:00 2001
Message-Id: <62821d2db6a1d60bc00c22d4da021999772bd2e4@dist-git>
From: Michal Privoznik <mprivozn@redhat.com>
Date: Tue, 8 Nov 2016 13:42:03 +0100
Subject: [PATCH] qemuBuildInterfaceCommandLine: Move hostdev handling a bit
 further

RHEL-7.3: https://bugzilla.redhat.com/show_bug.cgi?id=1366108
RHEL-7.3.z: https://bugzilla.redhat.com/show_bug.cgi?id=1392032

The idea is to have function that does some checking of the
arguments at its beginning and then have one big switch for all
the interface types it supports. Each one of them generating the
corresponding part of the command line.

Signed-off-by: Michal Privoznik <mprivozn@redhat.com>
(cherry picked from commit ec7f612a56ed8d3a2e83101b5cffcea3058fd30a)
Signed-off-by: Michal Privoznik <mprivozn@redhat.com>
---
 src/qemu/qemu_command.c                            | 13 ++++----
 .../qemuxml2argv-net-hostdev-fail.xml              | 39 ++++++++++++++++++++++
 tests/qemuxml2argvtest.c                           |  4 +++
 3 files changed, 49 insertions(+), 7 deletions(-)
 create mode 100644 tests/qemuxml2argvdata/qemuxml2argv-net-hostdev-fail.xml

diff --git a/src/qemu/qemu_command.c b/src/qemu/qemu_command.c
index 8d002ffb2..1e7d9209a 100644
--- a/src/qemu/qemu_command.c
+++ b/src/qemu/qemu_command.c
@@ -8007,13 +8007,6 @@ qemuBuildInterfaceCommandLine(virCommandPtr cmd,
     if (actualType == VIR_DOMAIN_NET_TYPE_VHOSTUSER)
         return qemuBuildVhostuserCommandLine(cmd, def, net, qemuCaps, bootindex);
 
-    if (actualType == VIR_DOMAIN_NET_TYPE_HOSTDEV) {
-        /* NET_TYPE_HOSTDEV devices are really hostdev devices, so
-         * their commandlines are constructed with other hostdevs.
-         */
-        return 0;
-    }
-
     /* Currently nothing besides TAP devices supports multiqueue. */
     if (net->driver.virtio.queues > 0 &&
         !(actualType == VIR_DOMAIN_NET_TYPE_NETWORK ||
@@ -8093,6 +8086,12 @@ qemuBuildInterfaceCommandLine(virCommandPtr cmd,
         if (qemuInterfaceEthernetConnect(def, driver, net,
                                        tapfd, tapfdSize) < 0)
             goto cleanup;
+    } else if (actualType == VIR_DOMAIN_NET_TYPE_HOSTDEV) {
+        /* NET_TYPE_HOSTDEV devices are really hostdev devices, so
+         * their commandlines are constructed with other hostdevs.
+         */
+        ret = 0;
+        goto cleanup;
     }
 
     /* For types whose implementations use a netdev on the host, add
diff --git a/tests/qemuxml2argvdata/qemuxml2argv-net-hostdev-fail.xml b/tests/qemuxml2argvdata/qemuxml2argv-net-hostdev-fail.xml
new file mode 100644
index 000000000..7807d793f
--- /dev/null
+++ b/tests/qemuxml2argvdata/qemuxml2argv-net-hostdev-fail.xml
@@ -0,0 +1,39 @@
+<domain type='qemu'>
+  <name>QEMUGuest1</name>
+  <uuid>c7a5fdbd-edaf-9455-926a-d65c16db1809</uuid>
+  <memory unit='KiB'>219136</memory>
+  <currentMemory unit='KiB'>219136</currentMemory>
+  <vcpu placement='static'>1</vcpu>
+  <os>
+    <type arch='i686' machine='pc'>hvm</type>
+    <boot dev='hd'/>
+  </os>
+  <clock offset='utc'/>
+  <on_poweroff>destroy</on_poweroff>
+  <on_reboot>restart</on_reboot>
+  <on_crash>destroy</on_crash>
+  <devices>
+    <emulator>/usr/bin/qemu</emulator>
+    <disk type='block' device='disk'>
+      <source dev='/dev/HostVG/QEMUGuest1'/>
+      <target dev='hda' bus='ide'/>
+      <address type='drive' controller='0' bus='0' target='0' unit='0'/>
+    </disk>
+    <controller type='usb' index='0'/>
+    <controller type='ide' index='0'/>
+    <controller type='pci' index='0' model='pci-root'/>
+    <interface type='hostdev' managed='yes'>
+      <mac address='00:11:22:33:44:55'/>
+      <source>
+        <address type='pci' domain='0x0000' bus='0x03' slot='0x07' function='0x1'/>
+      </source>
+      <model type='virtio'/>
+      <filterref filter='myfilter'/>
+      <backend tap='/dev/mytap'/>
+      <driver queues='4'/>
+    </interface>
+    <input type='mouse' bus='ps2'/>
+    <input type='keyboard' bus='ps2'/>
+    <memballoon model='virtio'/>
+  </devices>
+</domain>
diff --git a/tests/qemuxml2argvtest.c b/tests/qemuxml2argvtest.c
index 7ae7f5d66..d84c148df 100644
--- a/tests/qemuxml2argvtest.c
+++ b/tests/qemuxml2argvtest.c
@@ -1065,6 +1065,10 @@ mymain(void)
             QEMU_CAPS_DEVICE_VFIO_PCI, QEMU_CAPS_HOST_PCI_MULTIDOMAIN);
     DO_TEST_FAILURE("net-hostdev-vfio-multidomain",
                     QEMU_CAPS_NODEFCONFIG, QEMU_CAPS_DEVICE_VFIO_PCI);
+    DO_TEST_FAILURE("net-hostdev-fail",
+                    QEMU_CAPS_NODEFCONFIG,
+                    QEMU_CAPS_DEVICE_VFIO_PCI);
+
 
     DO_TEST("serial-vc", NONE);
     DO_TEST("serial-pty", NONE);
-- 
2.11.0

